core/src/test/java/edu/ucr/cs/riple/core/AnalysisModeTest.java

```diff
@@ -85,4 +86,27 @@ public void lowerBoundModeTest() {
         .enableDownstreamDependencyAnalysis(LOWER_BOUND)
         .start();
   }
+
+  @Test
+  public void upperBoundModeTest() {
+    coreTestHelper
+        .addExpectedReports(
+            new TReport(
+                new OnMethod("Foo.java", "test.target.Foo", "returnNullGood()"), -6, APPROVE),
+            // Resolves 6 errors locally but creates 1 error on downstream dependencies that cannot
+            // be resolved, one of the triggered fixes in the tree, triggers an error in downstream
+            // dependency, the overall effect is -6 + 1 + 1 = -4 in upper bound mode.
```

Comment:
Regarding upper bound and lower bound, please consider the assumptions below:
1. A root fix resolving error is `rf`.
2. `rf` triggers `f0`, `f1` and `f2` all targeting elements in the target module.
3. If `rf`,`f0`, `f1`, `f2` applied, it will decrease the number of errors on target by 6 -> effect = -6.
And below is the result of analysis of downstream dependencies on `dep1`, `dep2` and `dep3`:

`Cell[i][j]` is the number of triggered errors for fix `i` on dependency `j`.

|Fix\Dep|`dep1`|`dep2`|`dep3`| total |
|:------:|:-----:|:------:|:------:|:-----:|
|   `rf`   |   0    |      1     |      2    |     3   |
|  ` f0`  |  1     |      1     |      2    |     4   |
|   `f1`   |   1     |      9    |      0    |     10 |
|   `f2`  |   0    |      0    |      1     |     1   |

With the current approach, regardless of the mode configuration we can detect the set of triggered errors for each fix on each module individually.

For chain `rf` -> `f0`, `f1`, `f2`, since all fixes are applied we can't measure the effectiveness accurately since they may trigger the same error. 
In `upper bound` mode we assume all errors are distinct therefore the overall effect of the chains is:
-6 + 3 + 4 + 10 + 1 = 12.

In `lower bound` mode, we just consider the maximum number of errors as the effect of the whole chain on downstream dependencies:
-6 + 10 = 4.

Therefore upper bound and lower bound is regardless of the number of downstream dependencies.


And the part you mentioned regarding that filtering, that only filters triggered errors that are resolved due to a fix in the tree, it does not detect identical errors that two fixes in a tree can trigger. Please let me know your thoughts on this. Thank you.

Reply:
I had to think a bit to parse this explanation, but I think it tracks.

Let me see if I got it...

First, for a single fix (not a tree of fixes, just the individual `rf` or `f{i}` in isolation), there is no difficulty in computing the _precise_ effect of the fix on each dependency or the _precise_ effect of the fix across all dependencies (the **total** column above). The difficulty lies in combining the effects of all fixes in the tree, and this is true whether there are multiple dependencies or a single one. Correct?

So, going by the example from the test case, we have one dep and three fixes:

- `rf` : "Add `@Nullable` to `returnNullBad()`"
- `f0` : "Add `@Nullable` to argument 1 of `takeNullGood(...)`" (because of `bar()` in `Dep`)
- `f1` : "Add `@Nullable` to `getNull()`" (because that method returns the result of `returnNullBad()`)

So the table looks like:

|Fix\Dep|`dep`|explanation|
|:------:|:-----:|:-----:|
|   `rf`   |   1    |  2 errors generate, but one resolved by f1 |
|  ` f0`  |  0     |  |
|   `f1`   |   1     | body of `getNullDown()` |

Right?

And the effect on the target itself is removing 5 errors from `returnNullBad()` and one from `getNull()`, so -6+2=-4, which is both correct and what the "upper bound" method gives here.

Meanwhile, "lower bound" gives -6+1=-5, because it can't tell if the new error that shows when applying `f1` is a repeat of the one that shows up when applying `rf`.

And this:

> And the part you mentioned regarding that filtering, that only filters triggered errors that are resolved due to a fix in the tree, it does not detect identical errors that two fixes in a tree can trigger.

Is just the logic that filters the error from `foo.takeNullGood(foo.returnNullBad());` in `Dep::bar()`, because it can see a fix for that specific error in the tree (namely, it sees `f0`).

Reply:
Yes that is absolutely correct. 👍🏻